### ISSUE TITLE
DDL task queue is stuck

### ISSUE THREAD

**Issue by simPod** [ClickHouse/ClickHouse#15181]
`select * from system.zookeeper where path = '/clickhouse/task_queue/ddl';`

The ddl task queue is full (1000 items). It is not processing though so items are stuck there forever now, they are in the queue 4 days already.

Is there a way to find what's wrong and fix it? Error log shows nothing (no errors), there are no long running queries and replication queue is empty.

Every other query with `ON CLUSTER` clause now fails on timeout, I guess it's because it doesn't fit into the queue.

v20.9.2.20

**Comment by alesapin** [comment]
Maybe DDLWorker thread is dead for some reason? Can you dump traces with gdb using: `gdb -ex 'set pagination off' -ex "set logging file gdb.log" -ex 'set logging on' -ex 'backtrace' -ex 'thread apply all backtrace' -ex 'backtrace' --pid your_clickhouse_pid`. After that try to restart the server, DDLWorker should start and start processing queue.

**Comment by simPod** [comment]
Should dumping traces have any effect on it?

I'll ask admins to do continuous restart of the cluster.

**Comment by simPod** [comment]
Traces dump https://gist.github.com/simPod/578ef00a3874a199df1cd4dbd2fb13b9

**Comment by simPod** [comment]
Though restarting the replicas had no effect on items in the queue.

**Comment by simPod** [comment]
Ok it might have restarted ALTER query that runs now 🤔 I'll get back when it finishes.

**Comment by alesapin** [comment]
DDL worker seems to be alive:
```

Thread 211 (Thread 0x7f1e80ead700 (LWP 30444)):
#0  0x00007f1f2229b17f in pthread_cond_wait@@GLIBC_2.3.2 () from /lib/x86_64-linux-gnu/libpthread.so.0
#1  0x0000000018dfb36b in Poco::EventImpl::waitImpl() ()
#2  0x000000001588f277 in DB::DDLWorker::runMainThread() ()
#3  0x0000000015898817 in ThreadFromGlobalPool::ThreadFromGlobalPool<void (DB::DDLWorker::*)(), DB::DDLWorker*>(void (DB::DDLWorker::*&&)(), DB::DDLWorker*&&)::{lambda()#1}::operator()() const ()
#4  0x000000000e75d547 in ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) ()
#5  0x000000000e75bb83 in ?? ()
#6  0x00007f1f222954a4 in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#7  0x00007f1f21bcbd0f in clone () from /lib/x86_64-linux-gnu/libc.so.6
```

**Comment by simPod** [comment]
Yes, though the queue is still full. Is there any way to clear it?

**Comment by simPod** [comment]
It is processing, very slowly though 🤔

**Comment by alesapin** [comment]
Any ideas why it's processing so slow? Which queries you have in DDL queue?

**Comment by simPod** [comment]
Basically all are based on sync script: `CREATE TABLE IF NOT EXISTS ...` while it always exists as I'm not adding new tables.

CPU load is 30 % and error log is empty. There seems to be no reason.

**Comment by simPod** [comment]
I disabled the sync script and the queue is not processing anymore. It was probably somehow being overridden by trying to add new items to it.

**Comment by simPod** [comment]
Ok maybe I might have misinterpreted what I see in ZK. Select from `system.zookeeper` returns items in DDL queue but that should not be an issue. I checked nodes in ZK where there are `active` and `finished` nodes where `active` has 0 items in it and `finished` has `n of items = n of replicas`.

So maybe the queue is not stuck then.

**Comment by broken-ufa** [comment]
I have an identical issue. Restart of all nodes solves the problem, but in few weeks this problem return.
Is this problem solved in new versions?

**Comment by alexey-milovidov** [comment]
This problem is not solved.

DDL task queue can be processed slowly or even stuck if there are queries in this queue that are processed slowly or cannot be processed.

It can be introspected with `system.distributed_ddl_queue` table on every shard.

We are planning to implement parallel processing of independent queries in DDL queue. It depends on #15170 and will be solved in a separate task.

**Comment by FrankChen021** [comment]
Hi @alexey-milovidov May I know are you still working on the parallel execution of DDLs? Does it has something to do with current multiple threading DDL worker?
